### PR TITLE
Add init containers and affinity pod specs to validation check exlusion list

### LINKED ISSUE #45287: Scaling or deleting statefulset in 1.5 cluster with 1.6 kubectl fails
**BUG REPORT**

When running either `kubectl delete statefulset <statefulset name>` or `kubectl scale statefulset <statefulset name> --replicas=<count>` with kubectl 1.6.2 against a GKE cluster 1.5.6 it fails with an error like the following.

```
error: Scaling the resource failed with: StatefulSet.apps "<statefulset name>" is invalid: spec: Forbidden: updates to statefulset spec for fields other than 'replicas' are forbidden.; Current resource version 100392498
```

This happened the other way around as well with kubectl 1.5.x against a GKE 1.6.x cluster.

For scaling there's a working alernative by running `kubectl patch statefulsets <statefulset name> -p '{"spec":{"replicas":'<count>'}}'`. Unfortunately I haven't found a workaround yet for deleting the statefulset.

**Kubernetes version**:

```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.2", GitCommit:"477efc3cbe6a7effca06bd1452fa356e2201e1ee", GitTreeState:"clean", BuildDate:"2017-04-19T20:33:11Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"5", GitVersion:"v1.5.6", GitCommit:"114f8911f9597be669a747ab72787e0bd74c9359", GitTreeState:"clean", BuildDate:"2017-03-28T13:36:31Z", GoVersion:"go1.7.4", Compiler:"gc", Platform:"linux/amd64"}
```

**Environment**:
- **Google Container Engine 1.6.2 with cos image**:
- **OS** (e.g. from /etc/os-release):

```
BUILD_ID=9000.84.2
NAME="Container-Optimized OS"
GOOGLE_CRASH_ID=Lakitu
VERSION_ID=56
BUG_REPORT_URL=https://crbug.com/new
PRETTY_NAME="Container-Optimized OS from Google"
VERSION=56
GOOGLE_METRICS_PRODUCT_ID=26
HOME_URL="https://cloud.google.com/compute/docs/containers/vm-image/"
ID=cos
```

- **Kernel** (e.g. `uname -a`):

```
Linux gke-production-europe-we-auto-scaling-917da0af-zzzj 4.4.21+ #1 SMP Fri Feb 17 15:34:45 PST 2017 x86_64 Intel(
R) Xeon(R) CPU @ 2.50GHz GenuineIntel GNU/Linux
```

**What happened**:

- `kubectl delete statefulset <statefulset name>` failed with `updates to statefulset spec for fields other than 'replicas' are forbidden` error
- `kubectl scale statefulset <statefulset name> --replicas=<count>` failed with `updates to statefulset spec for fields other than 'replicas' are forbidden` error

**What you expected to happen**:

Both actions to succeed

**How to reproduce it** (as minimally and precisely as possible):

- Create a statefulset in a GKE 1.5.6 cluster
- Run the above commands with kubectl 1.6.2

**Anything else we need to know**:

Happens for kubectl 1.5.x with GKE 1.6.x as well.

### PR DESCRIPTION
**Pull request by crimsonfaith91**
**What this PR does / why we need it**:
Add init containers and affinity pod specs to exclusion list of validation check to ensure the check can pass for client and server of different versions.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #45287 

**Special notes for your reviewer**:

**Release note**:

```release-note
```

### CODE REVIEW

**Comment by liggitt** [comment]
looks like something went wrong with the source branch... make sure your validation branch consists of just your commit on top of current master.

also, it looks like the author preferences in git might not be right on your machine, it's reporting the author as root@jt-linux.svl.corp.google.com

https://github.com/crimsonfaith91/kubernetes/commit/4cc023d1e4eb1ac0df8b367f64d96cb14d674837.patch

**Comment by crimsonfaith91** [comment]
i think i pushed from the wrong commit. i will commit again. sorry!

**Comment by crimsonfaith91** [comment]
Perhaps a better way is to put the attributes into a list and go over them.

**Comment by foxish** [comment]
@k8s-bot ok to test

**Comment by foxish** [comment]
As we spoke offline, this needs unit tests to confirm that it works, and skew tests to confirm that it fixes https://github.com/kubernetes/kubernetes/issues/45287

**Comment by crimsonfaith91** [comment]
i wrote the unit tests, but was blocked by apiserver timeout. the apiserver works sometimes.

**Comment by foxish** [comment]
I'd suggest investigating why the timeout issue is occurring and root-causing it, because it seems to be recurring.

File: pkg/apis/apps/validation/validation.go [hunk 1]
```diff
@@ -105,19 +105,27 @@ func ValidateStatefulSetUpdate(statefulSet, oldStatefulSet *apps.StatefulSet) fi
...
 	statefulSet.Spec.Replicas = oldStatefulSet.Spec.Replicas
 
 	restoreContainers := statefulSet.Spec.Template.Spec.Containers
 	statefulSet.Spec.Template.Spec.Containers = oldStatefulSet.Spec.Template.Spec.Containers
 
+	restoreInitContainers := statefulSet.Spec.Template.Spec.InitContainers
+	statefulSet.Spec.Template.Spec.InitContainers = oldStatefulSet.Spec.Template.Spec.InitContainers
+
+	restoreAffinity := statefulSet.Spec.Template.Spec.Affinity
+	statefulSet.Spec.Template.Spec.Affinity = oldStatefulSet.Spec.Template.Spec.Affinity
```
Comment (liggitt) [inline pkg/apis/apps/validation/validation.go]: Are updates to these fields handled now? Are other changes required if these fields are allowed to mutate now?

Comment (crimsonfaith91) [inline pkg/apis/apps/validation/validation.go]: This change aims to solve validation error where kubectl v1.6 does not work with server v1.5. Perhaps a better fix is to patch replicas when scaling, or create a scale endpoint for statefulset. More information can be found here: https://github.com/kubernetes/kubernetes/issues/46005

**Comment by crimsonfaith91** [comment]
This will be fixed in future releases of Kubernetes. I am asked to not updating statefulset validation now.

**Pull request closed without merging**